File: android/app/src/main/java/com/kreativekoala/cleanup/billing/PaywallCoordinator.kt
package com.kreativekoala.cleanup.billing

import android.content.Context
import android.content.SharedPreferences
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject
import javax.inject.Singleton

// MARK: - Paywall Context

sealed class PaywallContext(val title: String, val subtitle: String) {
    data object Generic : PaywallContext(
        "Unlock Cleanup Pro",
        "Unlimited cleanups & duplicate removal"
    )

    data class AttemptDeleteDuplicates(val count: Int, val savedGB: Double) : PaywallContext(
        "Remove $count duplicates in one tap",
        "Unlimited photo cleanups included"
    )

    data class AttemptDeleteSimilar(val count: Int, val savedGB: Double) : PaywallContext(
        "Clean up $count similar photos",
        "Unlimited photo cleanups included"
    )

    data class AttemptDeleteScreenshots(val count: Int, val savedGB: Double) : PaywallContext(
        "Delete $count screenshots instantly",
        "Bulk delete screenshots anytime"
    )

    data class AttemptDeleteVideos(val count: Int, val savedGB: Double) : PaywallContext(
        "Free up ${"%.1f".format(savedGB)} GB from $count videos",
        "Compress & delete large videos"
    )

    data class AttemptCompressVideo(val savedGB: Double) : PaywallContext(
        "Save ${"%.1f".format(savedGB)} GB with compression",
        "Compress & delete large videos"
    )

    data class AttemptMergeContacts(val count: Int) : PaywallContext(
        "Merge $count duplicate contacts",
        "Keep your contacts organized"
    )

    data class AttemptArchive(val count: Int, val totalGB: Double) : PaywallContext(
        "Archive $count items (${"%.1f".format(totalGB)} GB)",
        "Keep files safe in the cloud instead of deleting"
    )
}

// MARK: - Pending Cleanup Action

sealed class PendingCleanupAction {
    data class DeleteDuplicates(val ids: Set<Long>) : PendingCleanupAction()
    data class DeleteSimilar(val ids: Set<Long>) : PendingCleanupAction()
    data class DeleteScreenshots(val ids: Set<Long>) : PendingCleanupAction()
    data class DeleteVideos(val ids: Set<Long>) : PendingCleanupAction()
    data class CompressVideo(val id: Long) : PendingCleanupAction()
    data class MergeContacts(val groupId: String) : PendingCleanupAction()
    data object OneTapCleanup : PendingCleanupAction()
}

// MARK: - Paywall Coordinator

/**
 * Port of iOS PaywallCoordinator.
 *
 * Manages paywall display, context-aware messaging, pending actions,
 * winback offers, and access control checks.
 */
@Singleton
class PaywallCoordinator @Inject constructor(
    @ApplicationContext context: Context,
    private val entitlementManager: EntitlementManager,
    private val freeUsageManager: FreeUsageManager
) {
    companion object {
        private const val PREFS_NAME = "paywall_prefs"
        private const val DISMISS_COUNT_KEY = "paywall_dismiss_count"
        private const val LAST_DISMISS_DATE_KEY = "paywall_last_dismiss_date"
        private const val WINBACK_SHOWN_DATE_KEY = "winback_last_shown_date"
    }

    private val prefs: SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private val _showPaywall = MutableStateFlow(false)
    val showPaywall: StateFlow<Boolean> = _showPaywall.asStateFlow()

    private val _paywallContext = MutableStateFlow<PaywallContext>(PaywallContext.Generic)
    val paywallContext: StateFlow<PaywallContext> = _paywallContext.asStateFlow()

    private val _pendingAction = MutableStateFlow<PendingCleanupAction?>(null)
    val pendingAction: StateFlow<PendingCleanupAction?> = _pendingAction.asStateFlow()

    private val _showArchivePaywall = MutableStateFlow(false)
    val showArchivePaywall: StateFlow<Boolean> = _showArchivePaywall.asStateFlow()

    private val _showWinbackOffer = MutableStateFlow(false)
    val showWinbackOffer: StateFlow<Boolean> = _showWinbackOffer.asStateFlow()

    private var paywallDismissCount: Int = prefs.getInt(DISMISS_COUNT_KEY, 0)

    // MARK: - Show Paywall

    fun showPaywall(
        context: PaywallContext,
        pendingAction: PendingCleanupAction? = null
    ) {
        _paywallContext.value = context
        _pendingAction.value = pendingAction
        _showPaywall.value = true
    }

    fun dismissPaywall() {
        _showPaywall.value = false
    }

    fun dismissArchivePaywall() {
        _showArchivePaywall.value = false
    }

    fun trackDismiss() {
        paywallDismissCount++
        prefs.edit()
            .putInt(DISMISS_COUNT_KEY, paywallDismissCount)
            .putLong(LAST_DISMISS_DATE_KEY, System.currentTimeMillis())
            .apply()
    }

    fun clearPendingAction() {
        _pendingAction.value = null
    }

    // MARK: - Access Checks

    enum class AccessResult {
        PRO,
        FREE_TIER,
        PAYWALL
    }

    fun checkAccess(
        context: PaywallContext,
        pendingAction: PendingCleanupAction? = null
    ): AccessResult {
        if (entitlementManager.isPro.value) return AccessResult.PRO

        val remaining = freeUsageManager.remainingFreeActions()
        if (remaining > 0) return AccessResult.FREE_TIER

        showPaywall(context, pendingAction)
        return AccessResult.PAYWALL
    }

    fun recordUsage(count: Int) {
        freeUsageManager.recordUsage(count)
    }

    fun checkAccessOrShowPaywall(
        context: PaywallContext,
        pendingAction: PendingCleanupAction? = null
    ): Boolean {
        if (entitlementManager.isPro.value) return true
        showPaywall(context, pendingAction)
        return false
    }

    // MARK: - Archive Access

    enum class ArchiveAccessResult {
        ALLOWED,
        NEEDS_PRO,
        NEEDS_ARCHIVE_SUBSCRIPTION
    }

    fun checkArchiveAccess(
        context: PaywallContext,
        pendingAction: PendingCleanupAction? = null
    ): ArchiveAccessResult {
        if (!entitlementManager.isPro.value) {
            showPaywall(PaywallContext.Generic, pendingAction)
            return ArchiveAccessResult.NEEDS_PRO
        }

        if (!entitlementManager.hasArchiveSubscription.value) {
            _showArchivePaywall.value = true
            _pendingAction.value = pendingAction
            return ArchiveAccessResult.NEEDS_ARCHIVE_SUBSCRIPTION
        }

        return ArchiveAccessResult.ALLOWED
    }

    // MARK: - Winback Logic

    fun checkWinbackEligibility() {
        if (entitlementManager.isPro.value) {
            _showWinbackOffer.value = false
            return
        }
        if (paywallDismissCount < 3) {
            _showWinbackOffer.value = false
            return
        }

        val lastDismiss = prefs.getLong(LAST_DISMISS_DATE_KEY, 0)
        if (lastDismiss == 0L) {
            _showWinbackOffer.value = false
            return
        }

        val daysSinceLastDismiss = (System.currentTimeMillis() - lastDismiss) / 86_400_000.0
        if (daysSinceLastDismiss < 1) {
            _showWinbackOffer.value = false
            return
        }

        val lastWinback = prefs.getLong(WINBACK_SHOWN_DATE_KEY, 0)
        if (lastWinback > 0) {
            val daysSinceWinback = (System.currentTimeMillis() - lastWinback) / 86_400_000.0
            if (daysSinceWinback < 1) {
                _showWinbackOffer.value = false
                return
            }
        }

        _showWinbackOffer.value = true
    }

    fun markWinbackShown() {
        prefs.edit().putLong(WINBACK_SHOWN_DATE_KEY, System.currentTimeMillis()).apply()
        _showWinbackOffer.value = false
    }

    val shouldShowNudgeBanner: Boolean
        get() = !entitlementManager.isPro.value && paywallDismissCount >= 2
}

File: ios/Cleanup/Subscription/PaywallCoordinator.swift
import SwiftUI

// MARK: - Paywall Context
enum PaywallContext {
    case generic
    case attemptCleanup(totalGB: Double, itemCount: Int)
    case attemptDeleteDuplicates(count: Int, savedGB: Double)
    case attemptDeleteSimilar(count: Int, savedGB: Double)
    case attemptDeleteScreenshots(count: Int, savedGB: Double)
    case attemptDeleteVideos(count: Int, savedGB: Double)
    case attemptCompressVideo(savedGB: Double)
    case attemptMergeContacts(count: Int)
    case attemptArchive(count: Int, totalGB: Double)
    case archiveStorageFull(usedGB: Double, limitGB: Double)

    var title: String {
        switch self {
        case .generic:
            return "Unlock Cleanup Pro"
        case .attemptCleanup(let totalGB, _):
            return "Free up \(String(format: "%.1f", totalGB)) GB instantly"
        case .attemptDeleteDuplicates(let count, _):
            return "Remove \(count) duplicates in one tap"
        case .attemptDeleteSimilar(let count, _):
            return "Clean up \(count) similar photos"
        case .attemptDeleteScreenshots(let count, _):
            return "Delete \(count) screenshots instantly"
        case .attemptDeleteVideos(let count, let savedGB):
            return "Free up \(String(format: "%.1f", savedGB)) GB from \(count) videos"
        case .attemptCompressVideo(let savedGB):
            return "Save \(String(format: "%.1f", savedGB)) GB with compression"
        case .attemptMergeContacts(let count):
            return "Merge \(count) duplicate contacts"
        case .attemptArchive(let count, let totalGB):
            return "Archive \(count) items (\(String(format: "%.1f", totalGB)) GB)"
        case .archiveStorageFull(let usedGB, let limitGB):
            return "Storage full (\(String(format: "%.1f", usedGB))/\(String(format: "%.0f", limitGB)) GB)"
        }
    }

    var subtitle: String {
        switch self {
        case .generic:
            return "Unlimited cleanups & duplicate removal"
        case .attemptCleanup, .attemptDeleteDuplicates, .attemptDeleteSimilar:
            return "Unlimited photo cleanups included"
        case .attemptDeleteScreenshots:
            return "Bulk delete screenshots anytime"
        case .attemptDeleteVideos, .attemptCompressVideo:
            return "Compress & delete large videos"
        case .attemptMergeContacts:
            return "Keep your contacts organized"
        case .attemptArchive:
            return "Keep files safe in the cloud instead of deleting"
        case .archiveStorageFull:
            return "Upgrade your archive plan for more space"
        }
    }

    var analyticsName: String {
        switch self {
        case .generic: return "generic"
        case .attemptCleanup: return "attempt_cleanup"
        case .attemptDeleteDuplicates: return "attempt_delete_duplicates"
        case .attemptDeleteSimilar: return "attempt_delete_similar"
        case .attemptDeleteScreenshots: return "attempt_delete_screenshots"
        case .attemptDeleteVideos: return "attempt_delete_videos"
        case .attemptCompressVideo: return "attempt_compress_video"
        case .attemptMergeContacts: return "attempt_merge_contacts"
        case .attemptArchive: return "attempt_archive"
        case .archiveStorageFull: return "archive_storage_full"
        }
    }
}

// MARK: - Pending Cleanup Action
enum PendingCleanupAction {
    case deleteDuplicates(ids: Set<String>)
    case deleteSimilar(ids: Set<String>)
    case deleteScreenshots(ids: Set<String>)
    case deleteVideos(ids: Set<String>)
    case compressVideo(id: String)
    case mergeContacts(groupId: UUID)
    case oneTapCleanup
    case archivePhotos(ids: Set<String>)
    case archiveVideos(ids: Set<String>)
    case archiveContacts(ids: [String])
}

// MARK: - Paywall Coordinator
@MainActor
class PaywallCoordinator: ObservableObject {
    static let shared = PaywallCoordinator()

    @Published var showPaywall: Bool = false
    @Published var paywallContext: PaywallContext = .generic
    @Published var pendingAction: PendingCleanupAction?
    @Published var paywallDismissCount: Int = 0
    @Published var showWinbackOffer: Bool = false

    // Archive paywall state
    @Published var showArchivePaywall: Bool = false
    @Published var archivePaywallContext: PaywallContext = .generic

    private let defaults = UserDefaults.standard
    private let dismissCountKey = "paywall_dismiss_count"
    private let lastDismissDateKey = "paywall_last_dismiss_date"
    private let winbackShownDateKey = "winback_last_shown_date"

    private init() {
        paywallDismissCount = defaults.integer(forKey: dismissCountKey)
    }

    // MARK: - Show Paywall
    func showPaywall(context: PaywallContext, pendingAction: PendingCleanupAction? = nil) {
        self.paywallContext = context
        self.pendingAction = pendingAction
        self.showPaywall = true
    }

    // MARK: - Dismiss Paywall
    func dismissPaywall() {
        showPaywall = false
    }

    // MARK: - Track Dismiss (called from sheet onDismiss when user didn't purchase)
    func trackDismiss() {
        paywallDismissCount += 1
        defaults.set(paywallDismissCount, forKey: dismissCountKey)
        defaults.set(Date().timeIntervalSince1970, forKey: lastDismissDateKey)
    }

    // MARK: - Clear Pending Action
    func clearPendingAction() {
        pendingAction = nil
    }

    // MARK: - Check Access and Show Paywall if Needed
    /// Returns .pro if user is pro, .freeTier if free deletions available, .paywall if paywall was shown
    enum AccessResult {
        case pro
        case freeTier(remaining: Int)
        case paywall
    }

    func checkAccess(
        category: FreeUsageManager.Category,
        requestedCount: Int,
        context: PaywallContext,
        pendingAction: PendingCleanupAction? = nil
    ) -> AccessResult {
        if EntitlementManager.shared.isPro {
            return .pro
        }

        let remaining = FreeUsageManager.shared.remainingFreeActions(for: category)
        if remaining > 0 {
            return .freeTier(remaining: remaining)
        }

        showPaywall(context: context, pendingAction: pendingAction)
        return .paywall
    }

    /// Returns true if user has access (isPro or free tier), false if paywall was shown
    func checkAccessOrShowPaywall(
        context: PaywallContext,
        pendingAction: PendingCleanupAction? = nil
    ) -> Bool {
        if EntitlementManager.shared.isPro {
            return true
        } else {
            showPaywall(context: context, pendingAction: pendingAction)
            return false
        }
    }

    // MARK: - Archive Access Check

    enum ArchiveAccessResult {
        case allowed
        case needsArchiveSubscription
        case quotaExceeded
    }

    func checkArchiveAccess(
        requiredBytes: Int64,
        context: PaywallContext,
        pendingAction: PendingCleanupAction? = nil
    ) -> ArchiveAccessResult {
        guard EntitlementManager.shared.hasArchiveSubscription else {
            showArchivePaywall = true
            archivePaywallContext = context
            self.pendingAction = pendingAction
            return .needsArchiveSubscription
        }

        // Quota is checked in confirmArchive() using local subscription tier
        return .allowed
    }

    // MARK: - Winback Offer Logic

    /// Call this on app foreground or after paywall dismiss to check if winback should show
    func checkWinbackEligibility() {
        guard !EntitlementManager.shared.isPro else {
            showWinbackOffer = false
            return
        }
        // Show winback after user has dismissed the paywall 3+ times
        guard paywallDismissCount >= 3 else {
            showWinbackOffer = false
            return
        }
        // Only show if last dismiss was at least 1 day ago (give them breathing room)
        let lastDismiss = defaults.double(forKey: lastDismissDateKey)
        guard lastDismiss > 0 else {
            showWinbackOffer = false
            return
        }
        let daysSinceLastDismiss = (Date().timeIntervalSince1970 - lastDismiss) / 86400
        guard daysSinceLastDismiss >= 1 else {
            showWinbackOffer = false
            return
        }
        // Don't show if we already showed winback today
        let lastWinbackShown = defaults.double(forKey: winbackShownDateKey)
        if lastWinbackShown > 0 {
            let daysSinceWinback = (Date().timeIntervalSince1970 - lastWinbackShown) / 86400
            guard daysSinceWinback >= 1 else {
                showWinbackOffer = false
                return
            }
        }
        showWinbackOffer = true
    }

    func markWinbackShown() {
        defaults.set(Date().timeIntervalSince1970, forKey: winbackShownDateKey)
        showWinbackOffer = false
    }

    /// Whether user qualifies for a nudge banner (lower threshold than full winback sheet)
    var shouldShowNudgeBanner: Bool {
        guard !EntitlementManager.shared.isPro else { return false }
        return paywallDismissCount >= 2
    }
}

// MARK: - Paywall Features
struct LocalPaywallFeature: Identifiable {
    let id = UUID()
    let icon: String
    let title: String
    let description: String

    static let allFeatures: [LocalPaywallFeature] = [
        LocalPaywallFeature(
            icon: "sparkles",
            title: "Unlimited Cleanups",
            description: "Clean your device as often as you want"
        ),
        LocalPaywallFeature(
            icon: "photo.on.rectangle.angled",
            title: "Delete Duplicates & Similar",
            description: "Find and remove duplicate photos instantly"
        ),
        LocalPaywallFeature(
            icon: "video.fill",
            title: "Large Video Management",
            description: "Compress or delete space-hogging videos"
        ),
        LocalPaywallFeature(
            icon: "lock.shield",
            title: "Private & Secure",
            description: "All scanning happens on your device"
        )
    ]
}
